### PR TITLE
Domains: Add titan email upsell page contents

### PR DESCRIPTION
**Pull request by runofthemillgeek**
Related to #54941 

#### Changes proposed in this Pull Request

Adds email upsell content to the new page setup in #54941 reusing the existing `<EmailProvidersComparison>` component. ~~Also incorporates the component design changes for the former component~~. The page can be viewed at:

```
http://wordpress.com/domains/add/:domain/email/:siteName
```

Changes I wanted to call out:

1. Refactored `EmailProvidersComparison` component to support props to inject text content for the Domains page.
2. ~~Extracted out `TitanNewMailbox` and `TitanNewMailboxList` components to fix styling issues since styles were previously clubbed inside `TitanAddMailboxes` component.~~ See #55249
3. ~~Modified `HeaderCakeBack` to support `iconPosition` for the _Skip_ button (not sure if we're going ahead with this)~~
4. ~~Removed the old trash button next to input field and added the _Remove this mailbox_ button adjacent to _Add another mailbox_ button, along with layout changes.~~ See #55529

#### Testing instructions
<details>
<summary>Click to expand</summary>

1. Run git checkout `add/titan-email-upsell-contents` and start dev server
2. Log into a WordPress.com account that has a site
3. Open the _Domains_ page (replace `yoursite` with a site slug that you have): `http://calypso.localhost:3000/domains/manage/yoursite.wordpress.com`
4. Click "Just search for a domain" button
5. Search for a domain or choose one of the suggestions by clicking _Select button_
5. Once the page loads, replace `google-workspace` in the domain with `email`. For example:
    If the URL shown is:
    ```
    http://calypso.localhost:3000/domains/add/example.in/google-workspace/example.wordpress.com
    ```
   
    replace it with:

    ```
    http://calypso.localhost:3000/domains/add/example.in/email/example.wordpress.com
    ```
6. Verify that the new route loads and renders without errors

Things to consider:

1. Need to verify the changes to `EmalProvidersComparison` and related components don't break/introduce issues in other areas like the _Emails_ section
2. ~~Layout changes adapt well in mobile/desktop layouts~~
</details>

#### Screenshots/Recording

<details>
<summary>Click to expand</summary>

##### Recording
_P.S. ignore the 1Password popover_

https://user-images.githubusercontent.com/2759499/129753672-b190ac28-7480-4f80-9ec2-b577368150b5.mp4

##### Screenshots
dp-flow__google__desktop
dp-flow__titan__desktop
dp-flow__mobile
</details>

---

#### TODO
- [x] Get feedback on the current approach
- [x] Fix the broken Google Workspace card
- [x] Fix the broken _Add Email_ functionality

### CODE REVIEW

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (a8ci18n) [inline client/my-sites/domains/email-providers-upsell/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **13** times:
`translate( 'Pick one of our flexible options to connect your domain with email and start getting emails @%(domainName)s today.' )` *ES Score: 12*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/pull/55183)

File: client/my-sites/email/titan-new-mailbox-list/index.jsx
Comment (a8ci18n) [inline client/my-sites/email/titan-new-mailbox-list/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Remove mailbox' )` *ES Score: 6*

**Comment by runofthemillgeek** [comment]
@daledupreez I have taken the approach of reusing the comparison component and injecting the strings we need for this flow. Let me know if this approach works. I also thought about copying this over to `my-sites/domains` and making the changes there but this seemed simpler/fits the case.

I'm having trouble at the moment with adding things from this stage to the checkout as well as getting the GSuite card to work because of the dependency on `domain` from the store state and the guards/validation we have around this. Would love to get your thoughts and suggestions for this.

https://github.com/Automattic/wp-calypso/blob/baf2f876d2b3ee87187cc5a9f2534969194f98e9/client/my-sites/email/email-providers-comparison/index.jsx#L184

cc @stephanethomas @olaseni

**Comment by daledupreez** [comment]
Whew! This is a nice big PR, @sangeeth96!

Given how much you're changing here, I think it may be worth splitting the PR up to simplify testing and verification of the changes.

I think it probably makes the most sense to create a PR that only refactors the mailbox list component. We can test that it's still working in its current location and can focus on some of the smaller changes you've added.

For the `domain` being wired in, I think we're going to need to tackle that fairly directly. The issue is that we don't have an already-registered domain to work with, but the component was originally built with that expectation.

I am thinking it may then make sense to use a slightly different approach for the component, likely doing something along these lines:
 * Add a flag/prop to `EmailProvidersComparison` to indicate that we don't have the domain object. It may be worth calling is something like `cartDomainName` to indicate that the domain is in the cart rather than already purchase.
 * Add a derived/calculated `domainName` prop to `EmailProvidersComparison`: this should be the value of `cartDomainName` when it is present, and fall back on `domain.name` otherwise -- it probably makes sense to compute this where we're currently identifying `domain` within `connect`.
 * Update the internals of `EmailProvidersComparison` in a few key ways:
    - Wherever we get the domain name, fall back on the new `domainName` prop.
    - Refactor any permissions checks using `domain` to skip the check when `cartDomainName` is specified
    - We probably need to refactor the price display logic to work without the `domain` object -- we can take that on as a later step so we know exactly what situation we're going to be in. We can definitely assume the purchase price is free for now, but we probably need to wire in the renewal pricing (I haven't examined that code closely)
 
I _think_ that should get us most of the way to getting this working. But I think it's going to be much faster to test and review if we take this on in smaller chunks.

**Comment by runofthemillgeek** [comment]
@daledupreez thanks for the detailed feedback, especially with respect to the "domain in the cart" bit. That could work.

I'll right away work to raise separate PRs for the refactored units. Can rebase this one once the refactoring PRs are tested and merged.

**Comment by runofthemillgeek** [comment]
@daledupreez have updated the code as per your feedback. Also removed the `HeaderCake` modifications I had made since I realized I can use `actionButton` prop there.

Let me know your thoughts and next steps.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [inline client/my-sites/domains/email-providers-upsell/index.jsx]: Nit: please add this before `domainAddNew`.
```suggestion
import CalypsoShoppingCartProvider from 'calypso/my-sites/checkout/calypso-shopping-cart-provider';
import { domainAddNew } from 'calypso/my-sites/domains/paths';
```

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [inline client/my-sites/domains/email-providers-upsell/index.jsx]: The wrapping `<>` and `</>` fragment is no longer necessary because we're only returning one child element, `CalypsoShoppingCartProvider`.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [inline client/my-sites/domains/email-providers-upsell/index.jsx]: We shouldn't be passing this as a function, but just as the translated string.
```suggestion
					headerTitle={ translate( 'Register %(domainName)s', { args: { domain }, comment } )}
```
We can do that because this code is specifying the domain of interest!
We may also save a bit on (re)renders because the value comparisons for dynamic functions may suggest that we have a new functions for each render cycle.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [inline client/my-sites/domains/email-providers-upsell/index.jsx]: As with `headerTitle` both of these props should just pass the result of `translate()` in rather than have it wrapped by a callback.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [reply client/my-sites/domains/email-providers-upsell/index.jsx]: Please adjust to remove the "from" in your string so we don't need re-translation.

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: Some comments:
1. Please move these new props to the `// Props passed to this component` section
2. As noted elsewhere, the incoming title/text values shouldn't be functions -- you can make them string|object 
prop types if needed. (I've illustrated that below.
3. It may be easier to switch the `is*Shown` props to be `hide*` so we can default to showing them with null incoming props, i.e. `hideEmailForwardingCard` and `hideEmailHeader`
4. It may be clearer to use `showSkipButton` instead of `isSkippable`, as the former is a bit clearer.
5. Can you document the injected/wired-in `domainName` prop?

```suggestion
		cartDomainName: PropTypes.string,
		headerTitle: PropTypes.oneOfType( [ PropTypes.string, PropTypes.object ] ),
		hideEmailForwardingCard: PropTypes.bool,
		hideEmailHeader: PropTypes.bool,
		isSkippable: PropTypes.bool,
		onSkipClick: PropTypes.func,
		promoHeaderDescription: PropTypes.oneOfType( [ PropTypes.string, PropTypes.object ] ),
		promoHeaderTitle: PropTypes.oneOfType( [ PropTypes.string, PropTypes.object ] ),
		selectedDomainName: PropTypes.string.isRequired,

		// Props injected via connect()
		currencyCode: PropTypes.string,
		currentRoute: PropTypes.string,
		domain: PropTypes.object,
		domainName: PropTypes.string,
		gSuiteProduct: PropTypes.object,
		isGSuiteSupported: PropTypes.bool.isRequired,
		productsList: PropTypes.object.isRequired,
```

File: client/my-sites/email/titan-add-mailboxes/index.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-add-mailboxes/index.jsx]: It looks like this was a mistake -- this results in the buttons being misaligned when you add multiple mailboxes to the list.

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx:701]: Does it also make sense to wire in an explicit `hasCartDomain` prop that we can guarantee will be a boolean? This will simplify a number of the code locations that need to make decisions based on the `domain` object.
```suggestion
		const domainName = ownProps.cartDomainName ?? domain.name;
		const hasCartDomain = Boolean( ownProps.cartDomainName );
```

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: This might be simpler to understand/follow if we build the domain list separately, especially because we should not be passing the `domains` prop when we don't have a domain object.
When we have `domain`, we should pass in `[ domain ]`, but when we don't have `domain` we should leave the prop `null`. Take a look at this line of code: https://github.com/Automattic/wp-calypso/blob/49ea2453c8d150e169d841b105ccd23a69e9a0b6/client/components/gsuite/gsuite-new-user-list/index.tsx#L76

```suggestion
		const domainList = domain ? [ domain ] : null;
		const formFields =
			hasCartDomain || domain ? (
				<FormFieldset>
					<GSuiteNewUserList
						extraValidation={ identityMap }
						domains={ domainList }
						onUsersChange={ this.onGoogleUsersChange }
						selectedDomainName={ selectedDomainName }
						users={ googleUsers }
						onReturnKeyPress={ this.onGoogleFormReturnKeyPress }
```

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: I'd rather have the `cartDomainName` check (or a check for my proposed `hasCartDomain` prop) occur here.
```suggestion
		const isEligibleForFreeTrial = hasCartDomain || domain?.titanMailSubscription?.isEligibleForIntroductoryOffer; // or Boolean( cartDomainName ) || ...
```

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: Why the change here? This will affect all uses of this component, which feels like it's out of scope for now.

File: client/my-sites/email/titan-new-mailbox-list/index.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-new-mailbox-list/index.jsx]: Can we push these additional elements and UX to a separate PR please? This will affect all locations where we're adding Titan mailboxes, and makes the UX inconsistent with adding G Suite/Google Workspace mailboxes.

I think we'll want to look at this in terms of whether it's a generally useful UX improvement, so I don't want this PR to get held up by that discussion.

File: client/my-sites/email/titan-new-mailbox/index.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-new-mailbox/index.jsx]: Same as for the list UX changes -- can we move these changes to a separate PR?

**Review by daledupreez — COMMENTED** [review]
Overall this is looking good, but we still have some work to do.

**Comment by daledupreez** [comment]
@saygunnyc, could you take a look at the UX for this upsell flow?

I am really concerned that it will be really easy to miss the small "Skip ->" option in the top right corner. Here is a screen shot of things as they stand.

Screenshot 2021-08-13 at 15 48 22

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/email-providers-comparison/index.jsx:701]: Certainly makes sense ✌️

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/email-providers-comparison/index.jsx]: The Figma file had this copy for the flow but I made an incorrect assumption that this will be the case for other screens. Should I just revert this back to what it was before (i.e _Add Professional Email_) or should I add a prop for the parent to customize this text?

File: client/my-sites/email/titan-new-mailbox-list/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/titan-new-mailbox-list/index.jsx]: Good call and I'm sorry about making the wrong assumption here too (that this change will be carried over to other places) 😞. Will work to fork out a separate PR for this ✌️

**Comment by saygunnyc** [comment]
> @saygunnyc, could you take a look at the UX for this upsell flow?
> 
> I am really concerned that it will be really easy to miss the small "Skip ->" option in the top right corner. Here is a screen shot of things as they stand.
>


Thank you for the ping on this one, @daledupreez !
Do we have any stats on how the previous skip affordance was performing over the Google Workspace upsell? Specifically, I'm wondering about the drop-off rate at this step to use as a base in assessing this updated 1st design iteration. Sharing for reference below:

 
Previous

To be perfectly honest, I think we should re-iterate on this and the email comparison page at some point– _since they are using shared components._ to reduce the amount of visual clutter that comes from the boxed approach of Calypso, which makes it hard to differentiate input fields from the containers at a first sight. We currently are challenging Calypso in many areas over Dotcom and once re-visiting this, I think we should challenge it over here too.

For specifically the skip affordance, I am finding the Google Upsell skip button harder to notice within the dispersed input fields that share almost the same styling. And in terms of placement, it's not a good approach to have customers go through the input fields before they see a skip button. 

Additionally, perhaps it's also worth considering the upsell to only render Professional Email? This page might become much simpler if that is a route we would like to go towards from a business standpoint as well. Also worth considering which plan owners should see this offer–_Given the Professional Email step is now withdrawn from signup, however with learnings about plan subscribers and the previous upsell was only rendering Google Workspace._

Only If we happen to notice this page is underperforming, the below exploration can be a quick swap solution in the interim however it's not ideal to 1. Repeat the skip CTA 2.Have the skip CTA below input fields another reason to re-visit this page.

Screen Shot 2021-08-13 at 3 57 15 PM

File: client/my-sites/email/titan-new-mailbox-list/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/titan-new-mailbox-list/index.jsx]: @daledupreez have moved these changes to #55529

**Comment by runofthemillgeek** [comment]
@saygunnyc the interim solution looks nice in my opinion. My thought process is that if the user has less screen real estate, scrolled to the bottom of this page and makes the decision to skip, would be easier to notice the labelled Skip button you added.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (daledupreez) [inline client/my-sites/domains/email-providers-upsell/index.jsx:37]: It may be worth wrapping this in `useCallback` or similar to memoize it, but I don't think we need that.

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: Based on the code that we're calling, I think that we should pass in `[]` when we don't have a `domain` as the function we're calling expects a `domain` object.
```suggestion
		const domains = domain ? [ domain ] : [];
```

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (daledupreez) [inline client/my-sites/email/email-providers-comparison/index.jsx]: Nit: here and elsewhere, it is often easier to read if the props are in alphabetical order.

**Review by daledupreez — CHANGES_REQUESTED** [review]
I have only one blocking comment, which is for the case where we're building an array containing a string instead of an object.

Other than that, everything is looking really good! I think I should be able to get this merged and deployed in the morning.

I think we should defer the skip button addition for a separate PR, as that will involve broader testing of these components.

My one other request is to update the PR description and screenshots to reflect your refactors -- this PR is now doing a lot less!

Thanks for sticking with all my reviews! 🙇

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/email-providers-comparison/index.jsx]: Done

File: client/my-sites/email/email-providers-comparison/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/email/email-providers-comparison/index.jsx]: Done.

File: client/my-sites/domains/email-providers-upsell/index.jsx
Comment (runofthemillgeek) [reply client/my-sites/domains/email-providers-upsell/index.jsx:37]: I think usually, this is [not necessary](https://kentcdodds.com/blog/usememo-and-usecallback).

**Comment by runofthemillgeek** [comment]
@daledupreez 

> I think we should defer the skip button addition for a separate PR, as that will involve broader testing of these components.

But then this view wouldn't have a skip button, right? Is that okay for this PR? If so, I will push a commit to remove the related changes.

> My one other request is to update the PR description and screenshots to reflect your refactors -- this PR is now doing a lot less!

Ah, forgot about that. Will update that.

> Thanks for sticking with all my reviews! 🙇

Thanks for the detailed feedback! ❤️

**Comment by daledupreez** [comment]
> > I think we should defer the skip button addition for a separate PR, as that will involve broader testing of these components.

> But then this view wouldn't have a skip button, right? Is that okay for this PR? If so, I will push a commit to remove the related changes.

Just to clarify my earlier comment, I was referring to the "not interested"/skip button within each form that was proposed in [this comment](https://github.com/Automattic/wp-calypso/pull/55183#issuecomment-898693269) by @saygunnyc. We definitely want the "Skip" button in the header section, but I think we can ship this PR without adding that work in as well.

**Comment by runofthemillgeek** [comment]
Thanks for the clarification @daledupreez! I've updated the PR description and screenshots/recording.

**Review by daledupreez — APPROVED** [review]
This LGTM! I forgot to approve yesterday when I reviewed.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6448550

Thank you @sangeeth96 for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6448550).

**Pull request merged**
